text/0517-io-os-reform.md

```diff
@@ -473,11 +642,284 @@ throughout IO, we can go on to explore the modules in detail.
+
+```rust
+// A reader that yields no bytes
+fn empty() -> Empty; // in theory just returns `impl Read`
+
+impl Read for Empty { ... }
+
+// A reader that yields `byte` repeatedly (generalizes today's ZeroReader)
+fn repeat(byte: u8) -> Repeat;
+
+impl Read for Repeat { ... }
+
+// A writer that ignores the bytes written to it (/dev/null)
+fn sink() -> Sink;
+
+impl Write for Sink { ... }
+
+// Copies all data from a `Read` to a `Write`, returning the amount of data
+// copied.
+pub fn copy<R, W>(r: &mut R, w: &mut W) -> Result<u64, Error>
+```
+
+Like `write_all`, the `copy` method will discard the amount of data already
+written on any error and also discard any partially read data on a `write`
+error. This method is intended to be a convenience and `write` should be used
+directly if this is not desirable.
+
+#### Seeking
+[Seeking]: #seeking
+
+The seeking infrastructure is largely the same as today's, except that
+`tell` is removed and the `seek` signature is refactored with more precise
+types:
+
+```rust
+pub trait Seek {
+    type Err;
+    // returns the new position after seeking
+    fn seek(&mut self, pos: SeekPos) -> Result<u64, Err>;
+}
+
+pub enum SeekPos {
+    FromStart(u64),
+    FromEnd(i64),
+    FromCur(i64),
+}
+```
+
+The old `tell` function can be regained via `seek(SeekPos::FromCur(0))`.
+
+#### Buffering
+[Buffering]: #buffering
+
+The current `Buffer` trait will be renamed to `BufRead` for
+clarity (and to open the door to `BufWrite` at some later
+point):
+
+```rust
+pub trait BufRead: Read {
+    fn fill_buf(&mut self) -> Result<&[u8], Error>;
+    fn consume(&mut self, amt: uint);
+
+    fn read_until(&mut self, byte: u8, buf: &mut Vec<u8>) -> Result<(), Error> { ... }
+    fn read_line(&mut self, buf: &mut String) -> Result<(), Error> { ... }
+}
+
+pub trait BufReadExt: BufRead {
+    // Split is an iterator over Result<Vec<u8>, Error>
+    fn split(&mut self, byte: u8) -> Split<Self> { ... }
+
+    // Lines is an iterator over Result<String, Error>
+    fn lines(&mut self) -> Lines<Self> { ... };
+
+    // Chars is an iterator over Result<char, Error>
+    fn chars(&mut self) -> Chars<Self> { ... }
+}
+```
+
+The `read_until` and `read_line` methods are changed to take explicit,
+mutable buffers, for similar reasons to `read_to_end`. (Note that
+buffer reuse is particularly common for `read_line`). These functions
+include the delimiters in the strings they produce, both for easy
+cross-platform compatibility (in the case of `read_line`) and for ease
+in copying data without loss (in particular, distinguishing whether
+the last line included a final delimiter).
+
+The `split` and `lines` methods provide iterator-based versions of
+`read_until` and `read_line`, and *do not* include the delimiter in
+their output. This matches conventions elsewhere (like `split` on
+strings) and is usually what you want when working with iterators.
+
+The `BufReader`, `BufWriter` and `BufStream` types stay
```

Comment:
Oops, forgot to call attention to that. But yes, the previously suggested names (like `BufferedReader`) seemed needlessly verbose by the standards of Rust's `std`.